features/netsocket/emac-drivers/TARGET_NXP_EMAC/TARGET_LPC546XX/lpc546xx_emac.cpp

```diff
@@ -0,0 +1,601 @@
+        if (consume_index >= ENET_TX_RING_LEN) {
+            consume_index = 0;
+        }
+    }
+    /* Restore access */
+    TXLockMutex.unlock();
+}
+
+/** \brief Ethernet receive interrupt handler
+ *
+ *  This function handles the receive interrupt of LPC546XX.
+ */
+void LPC546XX_EMAC::rx_isr()
+{
+    osThreadFlagsSet(thread, FLAG_RX);
+}
+
+void LPC546XX_EMAC::tx_isr()
+{
+    osThreadFlagsSet(thread, FLAG_TX);
+}
+
+void LPC546XX_EMAC::ethernet_callback(ENET_Type *base, enet_handle_t *handle, enet_event_t event, uint8_t channel, void *param)
+{
+    LPC546XX_EMAC *enet = static_cast<LPC546XX_EMAC *>(param);
+
+    switch (event)
+    {
+        case kENET_RxIntEvent:
+            enet->rx_isr();
+            break;
+        case kENET_TxIntEvent:
+            enet->tx_isr();
+            break;
+        default:
+            break;
+    }
+}
+
+/** \brief  Low level init of the MAC and PHY.
+ */
+bool LPC546XX_EMAC::low_level_init_successful()
+{
+    status_t status;
+    uint8_t i;
+    uint32_t refClock;
+    phy_speed_t phy_speed;
+    phy_duplex_t phy_duplex;
+    uint32_t phyAddr = 0;
+    bool link = false;
+    enet_config_t config;
+    uint32_t timeout = 0xFFFU;
+
+    lpc546xx_init_eth_hardware();
+
+    refClock = CLOCK_GetFreq(kCLOCK_CoreSysClk);
+
+    status = PHY_Init(ENET, phyAddr, refClock);
+    if (status != kStatus_Success) {
+        return false;
+    }
+
+    ENET_GetDefaultConfig(&config);
+    config.multiqueueCfg = NULL;
+
+    while ((!link) && timeout) {
+        PHY_GetLinkStatus(ENET, phyAddr, &link);
+        if (link) {
+            /* Get link information from PHY */
+            PHY_GetLinkSpeedDuplex(ENET, phyAddr, &phy_speed, &phy_duplex);
+
+            /* Change the MII speed and duplex for actual link status. */
+            config.miiSpeed = (enet_mii_speed_t)phy_speed;
+            config.miiDuplex = (enet_mii_duplex_t)phy_duplex;
+        }
+        timeout--;
+    }
+
+    if (!link) {
+        return false;
+    }
+
+    AT_NONCACHEABLE_SECTION_ALIGN(static enet_rx_bd_struct_t rx_desc_start_addr[ENET_RX_RING_LEN], ENET_BUFF_ALIGNMENT);
+    AT_NONCACHEABLE_SECTION_ALIGN(static enet_tx_bd_struct_t tx_desc_start_addr[ENET_TX_RING_LEN], ENET_BUFF_ALIGNMENT);
+
+    /* Create buffers for each receive BD */
+    for (i = 0; i < ENET_RX_RING_LEN; i++) {
+        rx_buff[i] = memory_manager->alloc_heap(ENET_ETH_MAX_FLEN, ENET_BUFF_ALIGNMENT);
+        if (NULL == rx_buff[i])
+            return false;
+
+        rx_ptr[i] = (uint32_t)memory_manager->get_ptr(rx_buff[i]);
+    }
+
+     /* prepare the buffer configuration. */
+    enet_buffer_config_t buffCfg = {
+        ENET_RX_RING_LEN,
+        ENET_TX_RING_LEN,
+        &tx_desc_start_addr[0],
+        &tx_desc_start_addr[0],
+        &rx_desc_start_addr[0],
+        &rx_desc_start_addr[ENET_RX_RING_LEN],
+        rx_ptr,
+        ENET_ALIGN(ENET_ETH_MAX_FLEN, ENET_BUFF_ALIGNMENT),
+    };
+
+    ENET_Init(ENET, &config, hwaddr, refClock);
```

Comment:
On the new emac interface power_up() mac address is not yet set at this point so it contains null string. At old lwip driver interface it was valid already at this point.  On new emac, mac is written by network stack to emac after the power_up() with void set_hwaddr() call. 

We are having test case failures (at least with IAR) where the device does not receive messages sent to its mac address. 

Could you check that mac address configuration works as expected. Just was testing purposes, we changed code so that mac is valid already here and the driver worked. But that cannot be a solution in new emac since mac written later using the set_hwaddr().

Reply:
Yes, it seems like you should be passing `NULL` here rather than `hwaddr`, so `ENET_Init` doesn't try programming the MAC address yet.

Not sure what the other test failures were.

Reply:
One option is to use the mbed_mac_address() call to set the address to hwaddr before the ENET_Init call.